File: solutions/nim/robot-name/1/robot_name.nim
# Dedicated to Shree DR.MDD
import random, sets

randomize()

type
  Robot* = object
    name*: string

const LettersCount = 2
const DigitsCount = 3

const AlphaSet = 'A'..'Z'
const NumSet = '0'..'9'

var takenNames = initHashSet[string]()

proc generateID: string =
  for i in 0 ..< LettersCount:
    result.add(rand(AlphaSet))
  for i in 0 ..< DigitsCount:
    result.add(rand(NumSet))

proc createUnique: string =
  while true:
    result = generateID()
    if result notin takenNames:
      takenNames.incl(result)
      break

proc makeRobot*: Robot =
  result.name = createUnique()

proc reset*(bot: var Robot): void =
  bot.name = createUnique()

File: solutions/php/robot-name/1/RobotName.php
<?php
// For my Shree DR.MDD

declare(strict_types=1);

class Robot
{
    private string $id;
    private static array $existingIds = [];

    public function __construct() {
        $this->id = self::createId();
    }

    private static function createId(): string {
        $newId = '';
        while(in_array($newId, self::$existingIds) || $newId === '') {
            $newId = self::randomChar() . self::randomChar() . rand(0, 9) . rand(0, 9) . rand(0, 9);
        }
        self::$existingIds[] = $newId;
        return $newId;
    }
    
    public function getName(): string
    {
        return $this->id;
    }

    private static function randomChar(): string {
        return chr(rand(65, 90));
    }

    public function reset(): void
    {
        $this->id = self::createId();
    }
}
